lib/identity_cache.rb

```diff
@@ -191,6 +193,21 @@ def fetch_multi(*keys)
       result
     end
 
+    def with_deferred_parent_expiration
+      raise NestedDeferredParentBlockError if Thread.current[:deferred_parent_expiration]
```

Comment:
The ergonomics of disallowing nested use make it unpredictable to compose. We could instead allow nested use with the outermost usage doing the actual invalidations.
```
def with_deferred_parent_expiration
  if Thread.current[:deferred_parent_expiration]
    return yield
  end

  Thread.current[:deferred_parent_expiration] = true
  Thread.current[:parent_records_for_cache_expiry] = Set.new

  result = yield

  Thread.current[:deferred_parent_expiration] = nil
  Thread.current[:parent_records_for_cache_expiry].each(&:expire_primary_index)
  result
ensure
  Thread.current[:deferred_parent_expiration] = nil
  Thread.current[:parent_records_for_cache_expiry].clear
end
```
Also the return value of `with_deferred_parent_expiration` can be that of the `yield` block passthru.

Reply:
Hmm, perhaps I'm misunderstanding but, in the code above, `return yield` will lead to unintended side effects. I believe the code will do the following:

```rb
IdentityCache.with_deferred_expiration do
  @parent.transaction do
    @parent.children.destroy_all # Thread.current[:parent_records_for_cache_expiry] => [@parent]
    IdentityCache.with_deferred_expiration do # The `ensure` block runs when we `return yield`, so Thread.current[:parent_records_for_cache_expiry] => `nil`
      @unrelated_record.update # Thread.current[:parent_records_for_cache_expiry] => [<parent of @unrelated_record>]
    end
  end
end
```

I'm fairly sure `ensure` always runs, even when returning early, which means that `@parent` wouldn't be expired here, even though its associated `children` records were destroyed and had it queued for expiry.

> may not need a separate boolean with :parent_records_for_cache_expiry being nil before/after use

I believe we will need some "global" marker to inform the `ParentModelExpiration` class when to alter its behaviour as this is not default behaviour and we don't want to alter behaviour in all cases.

> return value of with_deferred_parent_expiration can be that of the yield block passthru.

To my shame, I missed this and it definitely needed fixing, thanks for flagging this! I have amended, rebased and pushed the changed behaviour.

> since Thread.current is accessible by name from any method it would be good to use a prefix for the key, e.g. idc_parent_records_for_cache_expiry

Great suggestion, have prefixed the `Thread.current` hash keys to prefix `idc_`.

Reply:
That's right about early `nil`ing with nested usage, we only want the `ensure` behaviour on the outermost one. We can get that by using a nest level counter, incremented on each entry of `with_deferred_parent_expiration` decremented for each exit and only do the `nil`ing when zero.